### PR TITLE
Apply container image config

### PR DESCRIPTION
**Pull request by Random-Liu**
Based on #41, #46 and #48.

This PR pulls container image and applies image config.

Only the last commit is new. Will rebase the PR and previous PRs soon.

### CODE REVIEW

**Comment by Random-Liu** [comment]
I've run all CRI validation test. Only the container log and `execSync` test failed, which was expected.

```
Running Suite: E2ECRI Suite
===========================
Random Seed: 1495696299 - Will randomize all specs
Will run 7 of 36 specs

S
------------------------------
[k8s.io] Container runtime should support basic operations on container 
  runtime should support starting container [Conformance]
  /home/lantaol/workspace/src/github.com/kubernetes-incubator/cri-tools/pkg/validate/container.go:98
[BeforeEach] [k8s.io] Container
  /home/lantaol/workspace/src/github.com/kubernetes-incubator/cri-tools/pkg/framework/framework.go:50
[BeforeEach] [k8s.io] Container
  /home/lantaol/workspace/src/github.com/kubernetes-incubator/cri-tools/pkg/validate/container.go:66
[BeforeEach] runtime should support basic operations on container
  /home/lantaol/workspace/src/github.com/kubernetes-incubator/cri-tools/pkg/validate/container.go:74
[It] runtime should support starting container [Conformance]
  /home/lantaol/workspace/src/github.com/kubernetes-incubator/cri-tools/pkg/validate/container.go:98
STEP: create container
STEP: Get image status
STEP: Pull image : busybox:1.26
STEP: Create container.
May 25 07:11:42.486: INFO: Created container "f5f5ab0779ab5e141e863efdc313b03079bc08d56de4fe2ba63cb59a32a5d819"

STEP: test start container
STEP: Start container for containerID: f5f5ab0779ab5e141e863efdc313b03079bc08d56de4fe2ba63cb59a32a5d819
May 25 07:11:42.599: INFO: Started container "f5f5ab0779ab5e141e863efdc313b03079bc08d56de4fe2ba63cb59a32a5d819"

STEP: Get container status for containerID: f5f5ab0779ab5e141e863efdc313b03079bc08d56de4fe2ba63cb59a32a5d819
[AfterEach] runtime should support basic operations on container
  /home/lantaol/workspace/src/github.com/kubernetes-incubator/cri-tools/pkg/validate/container.go:81
STEP: stop PodSandbox
STEP: delete PodSandbox
[AfterEach] [k8s.io] Container
  /home/lantaol/workspace/src/github.com/kubernetes-incubator/cri-tools/pkg/framework/framework.go:51
•SSS
------------------------------
[k8s.io] Container runtime should support log 
  runtime should support starting container with log [Conformance]
  /home/lantaol/workspace/src/github.com/kubernetes-incubator/cri-tools/pkg/validate/container.go:203
[BeforeEach] [k8s.io] Container
  /home/lantaol/workspace/src/github.com/kubernetes-incubator/cri-tools/pkg/framework/framework.go:50
[BeforeEach] [k8s.io] Container
  /home/lantaol/workspace/src/github.com/kubernetes-incubator/cri-tools/pkg/validate/container.go:66
[BeforeEach] runtime should support log
  /home/lantaol/workspace/src/github.com/kubernetes-incubator/cri-tools/pkg/validate/container.go:175
STEP: create a PodSandbox with log directory
[It] runtime should support starting container with log [Conformance]
  /home/lantaol/workspace/src/github.com/kubernetes-incubator/cri-tools/pkg/validate/container.go:203
STEP: create container with log
STEP: create a container with log and name
STEP: Get image status
STEP: Create container.
May 25 07:11:42.981: INFO: Created container "c12556a046e90f1b7499414d4fcbe821b6bb5de94201ed24522d587a48434c1c"

STEP: start container with log
STEP: Start container for containerID: c12556a046e90f1b7499414d4fcbe821b6bb5de94201ed24522d587a48434c1c
May 25 07:11:43.110: INFO: Started container "c12556a046e90f1b7499414d4fcbe821b6bb5de94201ed24522d587a48434c1c"

STEP: Get container status for containerID: c12556a046e90f1b7499414d4fcbe821b6bb5de94201ed24522d587a48434c1c
STEP: check the log context
STEP: verify log contents
May 25 07:11:43.111: INFO: Unexpected error occurred: open /tmp/podLogTest367317090/PodSandbox-with-log-directory-67a8ac64-4119-11e7-a07f-42010af00003/container-with-log-test-67cfd4b5-4119-11e7-a07f-42010af00003.log: no such file or directory
[AfterEach] runtime should support log
  /home/lantaol/workspace/src/github.com/kubernetes-incubator/cri-tools/pkg/validate/container.go:184
STEP: stop PodSandbox
STEP: delete PodSandbox
STEP: clean up the TempDir
[AfterEach] [k8s.io] Container
  /home/lantaol/workspace/src/github.com/kubernetes-incubator/cri-tools/pkg/framework/framework.go:51

• Failure [0.536 seconds]
[k8s.io] Container
/home/lantaol/workspace/src/github.com/kubernetes-incubator/cri-tools/pkg/framework/framework.go:72
  runtime should support log
  /home/lantaol/workspace/src/github.com/kubernetes-incubator/cri-tools/pkg/validate/container.go:204
    runtime should support starting container with log [Conformance] [It]
    /home/lantaol/workspace/src/github.com/kubernetes-incubator/cri-tools/pkg/validate/container.go:203

    failed to open log file: open /tmp/podLogTest367317090/PodSandbox-with-log-directory-67a8ac64-4119-11e7-a07f-42010af00003/container-with-log-test-67cfd4b5-4119-11e7-a07f-42010af00003.log: no such file or directory
    Expected error:
        <*os.PathError | 0xc4203252c0>: {
            Op: "open",
            Path: "/tmp/podLogTest367317090/PodSandbox-with-log-directory-67a8ac64-4119-11e7-a07f-42010af00003/container-with-log-test-67cfd4b5-4119-11e7-a07f-42010af00003.log",
            Err: 0x2,
        }
        open /tmp/podLogTest367317090/PodSandbox-with-log-directory-67a8ac64-4119-11e7-a07f-42010af00003/container-with-log-test-67cfd4b5-4119-11e7-a07f-42010af00003.log: no such file or directory
    not to have occurred

    /home/lantaol/workspace/src/github.com/kubernetes-incubator/cri-tools/pkg/validate/container.go:453
------------------------------
S
------------------------------
[k8s.io] Container runtime should support basic operations on container 
  runtime should support creating container [Conformance]
  /home/lantaol/workspace/src/github.com/kubernetes-incubator/cri-tools/pkg/validate/container.go:90
[BeforeEach] [k8s.io] Container
  /home/lantaol/workspace/src/github.com/kubernetes-incubator/cri-tools/pkg/framework/framework.go:50
[BeforeEach] [k8s.io] Container
  /home/lantaol/workspace/src/github.com/kubernetes-incubator/cri-tools/pkg/validate/container.go:66
[BeforeEach] runtime should support basic operations on container
  /home/lantaol/workspace/src/github.com/kubernetes-incubator/cri-tools/pkg/validate/container.go:74
[It] runtime should support creating container [Conformance]
  /home/lantaol/workspace/src/github.com/kubernetes-incubator/cri-tools/pkg/validate/container.go:90
STEP: test create a default container
STEP: Get image status
STEP: Create container.
May 25 07:11:43.505: INFO: Created container "777831fdf267bc74bd155289852633201e886fc41d862d7ef9ff66ac70fa0e0f"

STEP: Get container status for containerID: 777831fdf267bc74bd155289852633201e886fc41d862d7ef9ff66ac70fa0e0f
STEP: test list container
STEP: List containers for containerID: 777831fdf267bc74bd155289852633201e886fc41d862d7ef9ff66ac70fa0e0f
[AfterEach] runtime should support basic operations on container
  /home/lantaol/workspace/src/github.com/kubernetes-incubator/cri-tools/pkg/validate/container.go:81
STEP: stop PodSandbox
STEP: delete PodSandbox
[AfterEach] [k8s.io] Container
  /home/lantaol/workspace/src/github.com/kubernetes-incubator/cri-tools/pkg/framework/framework.go:51
•SS
------------------------------
[k8s.io] Container runtime should support basic operations on container 
  runtime should support stopping container [Conformance]
  /home/lantaol/workspace/src/github.com/kubernetes-incubator/cri-tools/pkg/validate/container.go:109
[BeforeEach] [k8s.io] Container
  /home/lantaol/workspace/src/github.com/kubernetes-incubator/cri-tools/pkg/framework/framework.go:50
[BeforeEach] [k8s.io] Container
  /home/lantaol/workspace/src/github.com/kubernetes-incubator/cri-tools/pkg/validate/container.go:66
[BeforeEach] runtime should support basic operations on container
  /home/lantaol/workspace/src/github.com/kubernetes-incubator/cri-tools/pkg/validate/container.go:74
[It] runtime should support stopping container [Conformance]
  /home/lantaol/workspace/src/github.com/kubernetes-incubator/cri-tools/pkg/validate/container.go:109
STEP: create container
STEP: Get image status
STEP: Create container.
May 25 07:11:43.908: INFO: Created container "13eb209a67a719601c9ac8fd473c1c277ae2b7c809d2052cd98fde9bb0f5d150"

STEP: start container
STEP: Start container for containerID: 13eb209a67a719601c9ac8fd473c1c277ae2b7c809d2052cd98fde9bb0f5d150
May 25 07:11:44.045: INFO: Started container "13eb209a67a719601c9ac8fd473c1c277ae2b7c809d2052cd98fde9bb0f5d150"

STEP: test stop container
STEP: Stop container for containerID: 13eb209a67a719601c9ac8fd473c1c277ae2b7c809d2052cd98fde9bb0f5d150
May 25 07:11:44.153: INFO: Stopped container "13eb209a67a719601c9ac8fd473c1c277ae2b7c809d2052cd98fde9bb0f5d150"

STEP: Get container status for containerID: 13eb209a67a719601c9ac8fd473c1c277ae2b7c809d2052cd98fde9bb0f5d150
[AfterEach] runtime should support basic operations on container
  /home/lantaol/workspace/src/github.com/kubernetes-incubator/cri-tools/pkg/validate/container.go:81
STEP: stop PodSandbox
STEP: delete PodSandbox
[AfterEach] [k8s.io] Container
  /home/lantaol/workspace/src/github.com/kubernetes-incubator/cri-tools/pkg/framework/framework.go:51
•SSSSSSS
------------------------------
[k8s.io] Container runtime should support basic operations on container 
  runtime should support execSync [Conformance]
  /home/lantaol/workspace/src/github.com/kubernetes-incubator/cri-tools/pkg/validate/container.go:132
[BeforeEach] [k8s.io] Container
  /home/lantaol/workspace/src/github.com/kubernetes-incubator/cri-tools/pkg/framework/framework.go:50
[BeforeEach] [k8s.io] Container
  /home/lantaol/workspace/src/github.com/kubernetes-incubator/cri-tools/pkg/validate/container.go:66
[BeforeEach] runtime should support basic operations on container
  /home/lantaol/workspace/src/github.com/kubernetes-incubator/cri-tools/pkg/validate/container.go:74
[It] runtime should support execSync [Conformance]
  /home/lantaol/workspace/src/github.com/kubernetes-incubator/cri-tools/pkg/validate/container.go:132
STEP: create container
STEP: Get image status
STEP: Create container.
May 25 07:11:44.540: INFO: Created container "81049279d7c23a1fa7ab217553cf2838b0e624e2e4ae7b5783f73275238efd0b"

STEP: start container
STEP: Start container for containerID: 81049279d7c23a1fa7ab217553cf2838b0e624e2e4ae7b5783f73275238efd0b
May 25 07:11:44.659: INFO: Started container "81049279d7c23a1fa7ab217553cf2838b0e624e2e4ae7b5783f73275238efd0b"

STEP: test execSync
STEP: verify execSync output
STEP: execSync for containerID: 81049279d7c23a1fa7ab217553cf2838b0e624e2e4ae7b5783f73275238efd0b
E0525 07:11:44.660212   31273 remote_runtime.go:301] ExecSync 81049279d7c23a1fa7ab217553cf2838b0e624e2e4ae7b5783f73275238efd0b 'echo hello' from runtime service failed: rpc error: code = 2 desc = not implemented
May 25 07:11:44.660: INFO: Unexpected error occurred: rpc error: code = 2 desc = not implemented
[AfterEach] runtime should support basic operations on container
  /home/lantaol/workspace/src/github.com/kubernetes-incubator/cri-tools/pkg/validate/container.go:81
STEP: stop PodSandbox
STEP: delete PodSandbox
[AfterEach] [k8s.io] Container
  /home/lantaol/workspace/src/github.com/kubernetes-incubator/cri-tools/pkg/framework/framework.go:51

• Failure [0.476 seconds]
[k8s.io] Container
/home/lantaol/workspace/src/github.com/kubernetes-incubator/cri-tools/pkg/framework/framework.go:72
  runtime should support basic operations on container
  /home/lantaol/workspace/src/github.com/kubernetes-incubator/cri-tools/pkg/validate/container.go:133
    runtime should support execSync [Conformance] [It]
    /home/lantaol/workspace/src/github.com/kubernetes-incubator/cri-tools/pkg/validate/container.go:132

    failed to execSync in container "81049279d7c23a1fa7ab217553cf2838b0e624e2e4ae7b5783f73275238efd0b"
    Expected error:
        <*grpc.rpcError | 0xc4203bef60>: {
            code: 2,
            desc: "not implemented",
        }
        rpc error: code = 2 desc = not implemented
    not to have occurred

    /home/lantaol/workspace/src/github.com/kubernetes-incubator/cri-tools/pkg/validate/container.go:358
------------------------------
S
------------------------------
[k8s.io] Container runtime should support adding volume and device [Conformance] 
  runtime should support starting container with volume [Conformance]
  /home/lantaol/workspace/src/github.com/kubernetes-incubator/cri-tools/pkg/validate/container.go:166
[BeforeEach] [k8s.io] Container
  /home/lantaol/workspace/src/github.com/kubernetes-incubator/cri-tools/pkg/framework/framework.go:50
[BeforeEach] [k8s.io] Container
  /home/lantaol/workspace/src/github.com/kubernetes-incubator/cri-tools/pkg/validate/container.go:66
[BeforeEach] runtime should support adding volume and device [Conformance]
  /home/lantaol/workspace/src/github.com/kubernetes-incubator/cri-tools/pkg/validate/container.go:141
[It] runtime should support starting container with volume [Conformance]
  /home/lantaol/workspace/src/github.com/kubernetes-incubator/cri-tools/pkg/validate/container.go:166
STEP: create host path and flag file
STEP: create container with volume
STEP: create a container with volume and name
STEP: Get image status
STEP: Create container.
May 25 07:11:44.970: INFO: Created container "fe951ba5bb46a33b3a4d452844b04d1b51827da6810d1d5eba74183cc35e577e"

STEP: test start container with volume
STEP: Start container for containerID: fe951ba5bb46a33b3a4d452844b04d1b51827da6810d1d5eba74183cc35e577e
May 25 07:11:45.111: INFO: Started container "fe951ba5bb46a33b3a4d452844b04d1b51827da6810d1d5eba74183cc35e577e"

STEP: Get container status for containerID: fe951ba5bb46a33b3a4d452844b04d1b51827da6810d1d5eba74183cc35e577e
STEP: test container exit code
STEP: Get container status for containerID: fe951ba5bb46a33b3a4d452844b04d1b51827da6810d1d5eba74183cc35e577e
[AfterEach] runtime should support adding volume and device [Conformance]
  /home/lantaol/workspace/src/github.com/kubernetes-incubator/cri-tools/pkg/validate/container.go:148
STEP: stop PodSandbox
STEP: delete PodSandbox
[AfterEach] [k8s.io] Container
  /home/lantaol/workspace/src/github.com/kubernetes-incubator/cri-tools/pkg/framework/framework.go:51
•SSSSSSSSSSSSSS
------------------------------
[k8s.io] Container runtime should support basic operations on container 
  runtime should support removing container [Conformance]
  /home/lantaol/workspace/src/github.com/kubernetes-incubator/cri-tools/pkg/validate/container.go:119
[BeforeEach] [k8s.io] Container
  /home/lantaol/workspace/src/github.com/kubernetes-incubator/cri-tools/pkg/framework/framework.go:50
[BeforeEach] [k8s.io] Container
  /home/lantaol/workspace/src/github.com/kubernetes-incubator/cri-tools/pkg/validate/container.go:66
[BeforeEach] runtime should support basic operations on container
  /home/lantaol/workspace/src/github.com/kubernetes-incubator/cri-tools/pkg/validate/container.go:74
[It] runtime should support removing container [Conformance]
  /home/lantaol/workspace/src/github.com/kubernetes-incubator/cri-tools/pkg/validate/container.go:119
STEP: create container
STEP: Get image status
STEP: Create container.
May 25 07:11:45.504: INFO: Created container "636699ff92f1543f5339ff8a026b296f6ef4dd7dcd33eb75a16ee88f877bb006"

STEP: test remove container
STEP: Remove container for containerID: 636699ff92f1543f5339ff8a026b296f6ef4dd7dcd33eb75a16ee88f877bb006
May 25 07:11:45.505: INFO: Removed container "636699ff92f1543f5339ff8a026b296f6ef4dd7dcd33eb75a16ee88f877bb006"

STEP: List containers for containerID: 636699ff92f1543f5339ff8a026b296f6ef4dd7dcd33eb75a16ee88f877bb006
[AfterEach] runtime should support basic operations on container
  /home/lantaol/workspace/src/github.com/kubernetes-incubator/cri-tools/pkg/validate/container.go:81
STEP: stop PodSandbox
STEP: delete PodSandbox
[AfterEach] [k8s.io] Container
  /home/lantaol/workspace/src/github.com/kubernetes-incubator/cri-tools/pkg/framework/framework.go:51
•

Summarizing 2 Failures:

[Fail] [k8s.io] Container runtime should support log [It] runtime should support starting container with log [Conformance] 
/home/lantaol/workspace/src/github.com/kubernetes-incubator/cri-tools/pkg/validate/container.go:453

[Fail] [k8s.io] Container runtime should support basic operations on container [It] runtime should support execSync [Conformance] 
/home/lantaol/workspace/src/github.com/kubernetes-incubator/cri-tools/pkg/validate/container.go:358

Ran 7 of 36 Specs in 5.877 seconds
FAIL! -- 5 Passed | 2 Failed | 0 Pending | 29 Skipped --- FAIL: TestE2ECRI (5.88s)
FAIL

Ginkgo ran 1 suite in 5.950683559s
Test Suite Failed
exit status 1
```

**Comment by Random-Liu** [comment]
@mikebrow @yujuhong PR rebased, ready for review.

File: cmd/cri-containerd/cri_containerd.go [hunk 1]
```diff
@@ -48,6 +48,9 @@ func main() {
 	if err != nil {
 		glog.Exitf("Failed to create CRI containerd service %+v: %v", o, err)
 	}
+	if err := service.Start(); err != nil {
```
Comment (yujuhong) [inline cmd/cri-containerd/cri_containerd.go]: Why is this in the PR?

File: pkg/server/container_create.go [hunk 2]
```diff
@@ -68,9 +69,23 @@ func (c *criContainerdService) CreateContainer(ctx context.Context, r *runtime.C
 		Config:    config,
 	}
 
-	// TODO(random-liu): [P0] Prepare container rootfs.
-
-	// TODO(random-liu): [P0] Set ImageRef in ContainerMetadata with image id.
+	// Prepare container image snapshot. For container, the image should have
+	// been pulled before creating the container, so do not ensure the image.
+	image := config.GetImage().GetImage()
+	imageMeta, err := c.ensureImageExists(ctx, image, false /*ensure*/)
```
Comment (yujuhong) [inline pkg/server/container_create.go]: Commented on #48. Would prefer breaking down the function.

File: pkg/server/container_start.go [hunk 3]
```diff
@@ -288,6 +290,27 @@ func (c *criContainerdService) generateContainerSpec(id string, sandboxPid uint3
 	return g.Spec(), nil
 }
 
+// setOCIProcessArgs sets process args. It returns error if the final arg list
+// is empty.
+func setOCIProcessArgs(g *generate.Generator, config *runtime.ContainerConfig, imageConfig *imagespec.ImageConfig) error {
+	command, args := config.GetCommand(), config.GetArgs()
+	// The following logic is migrated from https://github.com/moby/moby/blob/master/daemon/commit.go
+	// TODO(random-liu): Figure out whether the following logic is
```
Comment (yujuhong) [inline pkg/server/container_start.go]: is...?

File: pkg/server/container_start.go [hunk 4]
```diff
@@ -288,6 +290,27 @@ func (c *criContainerdService) generateContainerSpec(id string, sandboxPid uint3
...
+	if len(command) == 0 {
```
Comment (yujuhong) [inline pkg/server/container_start.go]: I don't entirely understand why the code below is organized this way...

File: pkg/server/sandbox_run.go [hunk 5]
```diff
@@ -300,3 +296,16 @@ func (c *criContainerdService) generateSandboxContainerSpec(id string, config *r
 
 	return g.Spec(), nil
 }
+
+// addImageEnvs adds environment variables from image config. It returns error if
+// an invalid environment variable is encountered.
+func addImageEnvs(g *generate.Generator, imageEnvs []string) error {
```
Comment (Random-Liu) [inline pkg/server/sandbox_run.go]: @yujuhong We don't have env in the sandbox container, right?

File: cmd/cri-containerd/cri_containerd.go (hunk 1 above)
Comment (Random-Liu) [inline cmd/cri-containerd/cri_containerd.go]: Found this when running cri validation container test.

I could send another PR to fix this.

File: pkg/server/container_start.go (hunk 3 above)
Comment (Random-Liu) [inline pkg/server/container_start.go]: Incomplete comment, will complete it.

File: pkg/server/container_start.go (hunk 4 above)
Comment (Random-Liu) [inline pkg/server/container_start.go]: This is the current behavior of docker, I don't fully understood it, either. However, there is no definition of how this should behave, so I just reuse the docker logic.

The TODO is to clearly define this and fix the logic here.

File: pkg/server/container_create.go (hunk 2 above)
Comment (Random-Liu) [inline pkg/server/container_create.go]: Will do.

File: cmd/cri-containerd/cri_containerd.go (hunk 1 above)
Comment (Random-Liu) [inline cmd/cri-containerd/cri_containerd.go]: Since travis is overloaded, it takes a long time to test a PR.

I'll split this into another commit.

Comment (Random-Liu) [inline cmd/cri-containerd/cri_containerd.go]: Done.

File: pkg/server/container_create.go (hunk 2 above)
Comment (Random-Liu) [inline pkg/server/container_create.go]: Done.

File: pkg/server/container_start.go (hunk 3 above)
Comment (Random-Liu) [inline pkg/server/container_start.go]: Done.

File: pkg/server/container_start.go (hunk 4 above)
Comment (Random-Liu) [inline pkg/server/container_start.go]: Done.

**Comment by Random-Liu** [comment]
@yujuhong Addressed comments.

**Comment by yujuhong** [comment]
LGTM

**Comment by Random-Liu** [comment]
Apply lgtm based on https://github.com/kubernetes-incubator/cri-containerd/pull/52#issuecomment-305053910, will rebase and merge this PR after test passes and dependency is merged.
